module/extension/Director/src/bits/compare_priority.cpp

```diff
@@ -0,0 +1,87 @@
+/*
+ * This file is part of the NUbots Codebase.
+ *
+ * The NUbots Codebase is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * The NUbots Codebase is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with the NUbots Codebase.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Copyright 2021 NUbots <nubots@nubots.net>
+ */
+
+#include <algorithm>
+#include <tuple>
+#include <vector>
+
+#include "../Director.hpp"
+
+#include "extension/Behaviour.hpp"
+
+namespace module::extension {
+
+    using ::extension::behaviour::commands::DirectorTask;
+
+    bool Director::compare_priority(const std::shared_ptr<const DirectorTask>& incumbent,
+                                    const std::shared_ptr<const DirectorTask>& challenger) {
+
+        // Compare to yourself and you get that you shouldn't change tasks
+        if (incumbent->requester_id == challenger->requester_id) {
+            return false;
+        }
+
+        // Function to get the priorities of the ancestors of this task
+        auto get_ancestor_priorities = [this](const std::shared_ptr<const DirectorTask>& task) {
+            // Loop up through the providers until we reach a point where a task was emitted by a non provider
+            std::vector<std::tuple<uint64_t, int, bool>> ancestors;
+            for (auto t = task; providers.count(t->requester_id) != 0;) {
```

Comment:
Since loops are possible in the provider tree, this could loop inifinitely. You should either add a visitor list to prevent this or remove the possibility of loops

Reply:
I'll add loop prevention earlier in the algorithm.
If it makes it here with a loop then crashing is the appropriate result.